src/config/app.ts

```diff
@@ -0,0 +1,328 @@
+    readonly exclude: JobToExclude[];
+    readonly ignorePhpPlatformRequirements: IgnorePhpPlatformRequirements;
+    readonly additionalComposerArguments: string[];
+}
+export interface Requirements {
+    readonly codeChecks: boolean;
+    readonly docLinting: boolean;
+}
+
+function gatherJobExclusions(config: ConfigurationFromFile): JobToExclude[] {
+    if (isExplicitChecksConfiguration(config)) {
+        return [];
+    }
+
+    if (isConfigurationContainingJobExclusions(config)) {
+        return config.exclude ?? [];
+    }
+
+    return [];
+}
+function discoverPhpVersionsForJob(job: JobDefinitionFromFile, appConfig: Config): InstallablePhpVersionType[] {
+    if (!isInstallableVersion(job.php ?? '')) {
+        return [appConfig.stablePhpVersion];
+    }
+
+    const phpFromJob = job.php as InstallablePhpVersionType;
+
+    if (isAnyPhpVersionType(phpFromJob)) {
+        return appConfig.versions;
+    }
+
+    if (isLowestPhpVersionType(phpFromJob)) {
+        return [appConfig.minimumPhpVersion];
+    }
+
+    if (isLatestPhpVersionType(phpFromJob)) {
+        return [appConfig.latestPhpVersion];
+    }
+
+    return [ phpFromJob ];
+}
+
+function discoverComposerDependenciesForJob(job: JobDefinitionFromFile): ComposerDependency[] {
+    const dependencyFromConfig = job.dependencies ?? WILDCARD_ALIAS;
+
+    if (isAnyComposerDependency(dependencyFromConfig)) {
+        return [ComposerDependency.LOWEST, ComposerDependency.LATEST];
+    }
+
+    return [dependencyFromConfig];
+}
+
+function discoverIgnorePhpPlatformRequirementForJobByVersion(job: JobDefinitionFromFile, phpVersion: InstallablePhpVersionType, appConfig: Config): boolean {
+    if (job.ignore_php_platform_requirement ?? false) {
+        return true;
+    }
+
+    return appConfig.ignorePhpPlatformRequirements[phpVersion] ?? false;
+}
+
+function discoverAdditionalComposerArgumentsForCheck(job: JobDefinitionFromFile, appConfig: Config): string[] {
+    return [... new Set([
+        ... appConfig.additionalComposerArguments ?? [],
+        ... job.additional_composer_arguments ?? []
+    ])];
+}
+
+function convertJobDefinitionFromFileToJobDefinition(phpVersion: InstallablePhpVersionType, composerDependency: ComposerDependency, job: JobDefinitionFromFile, appConfig: Config): JobDefinition {
+
+    return {
+        php: phpVersion,
+        phpExtensions: job.extensions ?? appConfig.phpExtensions,
+        command: job.command,
+        phpIni: job.ini ?? appConfig.phpIni,
+        composerDependency: composerDependency,
+        ignorePhpPlatformRequirement: discoverIgnorePhpPlatformRequirementForJobByVersion(job, phpVersion, appConfig),
+        additionalComposerArguments: discoverAdditionalComposerArgumentsForCheck(job, appConfig)
+    };
+}
+
+function convertJobFromFileToJobs(job: JobFromFile, appConfig: Config): Job[] {
+    const jobDefinition: JobDefinitionFromFile = job.job;
+    const composerDependencies = discoverComposerDependenciesForJob(jobDefinition);
```

Comment:
Have to google to see what "restraints" is translated to. 😂 

I do not have a meaning here, can change this if we want to rename the initial naming.
I tried to keep as much old stuff as possible since I do not want to totally rework this but first move it to typescript.

Reply:
restraint or constraint, just to imply that this is adding a restriction to the dependency resolution. I wasn't aware the name dependencies was part of the configuration which we don't want to change though, so I'll give way to others opinions here, but when reading through the code, I was expecting this to be a list of required packages.

Reply:
Well, as a `constraint` I would assume we are talking about these `~8.0.0` or `^7.4 || ^8.0`, but actually we are talking about `locked`, `latest` and `lowest` dependencies.

Reply:
I've renamed it to `dependencySet` where possible.
This might reflect the meaning a little bit better as this will tell composer which dependency set to install:
- `locked` dependencies from `composer.lock`
- `latest` dependencies which are supported by the projects constraints
- `lowest` dependencies which are supported by the projects constraints